File: backend/src/main/java/com/hgw/gestionale/repair/mapper/RepairMapper.java
package com.hgw.gestionale.repair.mapper;

import com.hgw.gestionale.customer.entity.Customer;
import com.hgw.gestionale.customer.mapper.CustomerMapper;
import com.hgw.gestionale.product.mapper.ProductMapper;
import com.hgw.gestionale.repair.dto.RepairResponse;
import com.hgw.gestionale.repair.entity.Repair;
import com.hgw.gestionale.repairdetails.mapper.RepairDetailsMapper;
import com.hgw.gestionale.statorepair.entity.StatoRepair;
import com.hgw.gestionale.statorepair.mapper.StatoRepairMapper;
import com.hgw.gestionale.statoriparazione.entity.StatoRiparazione;
import com.hgw.gestionale.statoriparazione.mapper.StatoRiparazioneMapper;

public final class RepairMapper {

    private RepairMapper() {}

    public static Repair toEntity(Customer customer) {
        return Repair.builder()
                .customer(customer)
                .build();
    }

    public static RepairResponse toResponse(Repair repair) {
        return new RepairResponse(
                repair.getId(),
                CustomerMapper.toResponse(repair.getCustomer()),
                ProductMapper.toResponse(repair.getProduct()),
                RepairDetailsMapper.toResponse(repair.getDetails()),
                repair.getStato() != null ? StatoRepairMapper.toResponse(repair.getStato()) : null,
                repair.getStatoRiparazione() != null ? StatoRiparazioneMapper.toResponse(repair.getStatoRiparazione()) : null,
                repair.getCostoTotale()
        );
    }

    public static void updateEntity(Repair repair, Customer customer,
                                    StatoRepair stato, StatoRiparazione statoRiparazione) {
        repair.setCustomer(customer);
        repair.setStato(stato);
        repair.setStatoRiparazione(statoRiparazione);
    }
}

File: backend-node/src/mapper/repair.mapper.ts
import { RepairResponse } from "../dto/repair.dto";
import { repairs } from "../schema/repairs";
import { customers } from "../schema/customers";
import { products } from "../schema/products";
import { models } from "../schema/models";
import { brands } from "../schema/brands";
import { colors } from "../schema/colors";
import { repairDetails } from "../schema/repairDetails";
import { CustomerMapper } from "./customer.mapper";
import { ProductMapper } from "./product.mapper";
import { RepairDetailsMapper } from "./repairDetails.mapper";
import { InterventionQuantitaResponse, RepairMessageResponse } from "../dto/repairDetails.dto";

type RepairEntity = typeof repairs.$inferSelect;
type CustomerEntity = typeof customers.$inferSelect;
type ProductEntity = typeof products.$inferSelect;
type ModelEntity = typeof models.$inferSelect;
type BrandEntity = typeof brands.$inferSelect;
type ColorEntity = typeof colors.$inferSelect;
type RepairDetailsEntity = typeof repairDetails.$inferSelect;

export class RepairMapper {

    static toResponse(
        repair: RepairEntity,
        customer: CustomerEntity,
        product: ProductEntity,
        model: ModelEntity,
        brand: BrandEntity,
        color: ColorEntity,
        details: RepairDetailsEntity,
        interventi: InterventionQuantitaResponse[],
        messaggi: RepairMessageResponse[]
    ): RepairResponse {
        return {
            id: repair.id,
            customer: CustomerMapper.toResponse(customer),
            product: ProductMapper.toResponse(product, model, brand, color),
            details: RepairDetailsMapper.toResponse(details, interventi, messaggi),
            stato: repair.stato ?? undefined,
            statoRiparazione: repair.statoRiparazione ?? undefined,
            costoTotale: repair.costoTotale ? parseFloat(repair.costoTotale) : undefined,
            createdAt: repair.createdAt.toISOString(),
        };
    }

}
